overlord/ifacestate/apparmorprompting/common/common.go

```diff
@@ -336,18 +336,337 @@ func abstractPermissionsToAppArmorFilePermissions(iface string, permissions []st
+	allowablePathPattern       = fmt.Sprintf(`^/%s*$`, safePathChar)
+	allowablePathPatternRegexp = regexp.MustCompile(allowablePathPattern)
+)
+
+// Expands all groups in the given path pattern. Groups are enclosed by '{' '}'.
+// Returns a list of all the expanded path patterns, or an error if the given
+// pattern is invalid.
+func ExpandPathPattern(pattern string) ([]string, error) {
+	patternLen := len(pattern)
+	expanded := defaultPrefixes
+	currLiteralStart := 0
+	for i := 0; i < patternLen; i++ {
+		if pattern[i] == '\\' {
+			i += 1
+			if i == patternLen {
+				return nil, fmt.Errorf(`invalid path pattern: trailing unescaped '\' character: %q`, pattern)
+			}
+			continue
+		}
+		if pattern[i] == '}' {
+			return nil, fmt.Errorf(`invalid path pattern: unmatched '}' character: %q`, pattern)
+		}
+		if pattern[i] != '{' {
+			continue
+		}
+		infix := pattern[currLiteralStart:i]
+		groupExpanded, groupEnd, err := expandPathPatternFromIndex(pattern, i+1, patternLen)
+		if err != nil {
+			return nil, err
+		}
+		newExpanded := make([]string, 0, len(expanded)*len(groupExpanded))
+		for _, prefix := range expanded {
+			for _, suffix := range groupExpanded {
+				newExpanded = append(newExpanded, prefix+infix+suffix)
+			}
+		}
+		expanded = newExpanded
+		currLiteralStart = groupEnd + 1
+		i = groupEnd // let for loop increment to index after '}'
+	}
+	if len(expanded) == 1 && expanded[0] == "" {
+		// Didn't expand any groups, so return whole pattern.
+		return []string{cleanPattern(pattern)}, nil
+	}
+	// Append trailing literal string, if any, to all previously-expanded
+	// patterns, and clean the resulting patterns.
+	alreadySeen := make(map[string]bool, len(expanded))
+	newExpanded := make([]string, 0, len(expanded))
+	suffix := pattern[currLiteralStart:patternLen]
+	for _, prefix := range expanded {
+		final := cleanPattern(prefix + suffix)
+		if alreadySeen[final] {
+			continue
+		}
+		alreadySeen[final] = true
+		newExpanded = append(newExpanded, final)
+	}
+	expanded = newExpanded
+	return expanded, nil
+}
 
-// Checks that the given path pattern is valid.  Returns nil if so, otherwise
+var defaultPrefixes = []string{""}
+
+// Expands the contents of a group in the given path pattern, beginning at the
+// given index, until a '}' is seen. The given index should be the index of the
+// first character after the opening '{' of the group. Returns the list of
+// expanded strings, as well as the index of the closing '}' character.
+// Whenever a ',' character is encountered, cuts off the current sub-pattern
+// and begins a new one. Any '\'-escaped '{', ',', and '}' characters are
+// treated as literals. If the pattern terminates before a non-escaped '}' is
+// seen, returns an error.
+func expandPathPatternFromIndex(pattern string, index, patternLen int) (expanded []string, end int, err error) {
+	// Record total list of expanded patterns, to which other lists are appended
+	expanded = []string{}
```

Comment:
Named returns are good for documentation, but it's usually better to shadow them---for a moment I thought there was a global called 'expanded'!

Also, the initial value of this `expanded` is slightly different from the one in the above function `[]string{""}`

Reply:
Yes, within a group vs between groups I'm handling things a bit differently. Every time a comma (or '}') is seen, all the expanded patterns for the current group option are appended to the previous ones, whereas between groups, we combine the previosly-expanded and newly-expanded patterns via a Cartesian product of sorts.

I agree, the names I used are inconsistent. I'm going to rename `expanded` in `ExpandPathPattern` to `expandedPrefixes`. But I'm not sure what you mean in this case... I'm returning a named value (or `nil`) for every return statement, but do you mean that expanded should be `var expanded []string` ?

Reply:
Oh no I mean `expanded := []string{}` should work unless I'm mistaken